plugins/synaptics-rmi/fu-synaptics-rmi-device.c

```diff
@@ -0,0 +1,1070 @@
+	g_autofree gchar *instance_id = NULL;
+	g_autoptr(GByteArray) f01_basic = NULL;
+	g_autoptr(GByteArray) f01_db = NULL;
+	g_autoptr(GByteArray) f01_product_id = NULL;
+	g_autoptr(GByteArray) f34_ctrl = NULL;
+
+	/* read basic device information */
+	if (!fu_synaptics_rmi_device_set_rma_page (self, 0x00, error))
+		return FALSE;
+
+	/* read PDT */
+	if (!fu_synaptics_rmi_device_scan_pdt (self, error))
+		return FALSE;
+	priv->f01 = fu_synaptics_rmi_device_get_function (self, 0x01, error);
+	if (priv->f01 == NULL)
+		return FALSE;
+	addr = priv->f01->query_base;
+	f01_basic = fu_synaptics_rmi_device_read (self, addr, RMI_DEVICE_F01_BASIC_QUERY_LEN, error);
+	if (f01_basic == NULL) {
+		g_prefix_error (error, "failed to read the basic query: ");
+		return FALSE;
+	}
+	priv->manufacturer_id = f01_basic->data[0];
+	priv->has_lts = (f01_basic->data[1] & RMI_DEVICE_F01_QRY1_HAS_LTS) > 0;
+	priv->has_sensor_id = (f01_basic->data[1] & RMI_DEVICE_F01_QRY1_HAS_SENSOR_ID) > 0;
+	priv->has_query42 = (f01_basic->data[1] & RMI_DEVICE_F01_QRY1_HAS_PROPS_2) > 0;
+
+	/* use the product ID as an instance ID */
+	addr += 11;
+	f01_product_id = fu_synaptics_rmi_device_read (self, addr, RMI_PRODUCT_ID_LENGTH, error);
+	if (f01_product_id == NULL) {
+		g_prefix_error (error, "failed to read the product id: ");
+		return FALSE;
+	}
+	board_id = g_strndup ((const gchar *) f01_product_id->data, f01_product_id->len);
+	instance_id = g_strdup_printf ("SYNAPTICS_RMI\\%s", board_id);
+	fu_device_add_instance_id (device, instance_id);
+
+	/* skip */
+	prod_info_addr = addr + 6;
+	addr += 10;
+	if (priv->has_lts)
+		addr++;
+
+	/* get sensor ID */
+	if (priv->has_sensor_id) {
+		g_autoptr(GByteArray) f01_tmp = NULL;
+		f01_tmp = fu_synaptics_rmi_device_read (self, addr++, 1, error);
+		if (f01_tmp == NULL) {
+			g_prefix_error (error, "failed to read the sensor id: ");
+			return FALSE;
+		}
+		priv->sensor_id = f01_tmp->data[0];
+	}
+
+	/* skip */
+	if (priv->has_lts)
+		addr += RMI_DEVICE_F01_LTS_RESERVED_SIZE;
+
+	/* read package ids */
+	if (priv->has_query42) {
+		g_autoptr(GByteArray) f01_tmp = NULL;
+		f01_tmp = fu_synaptics_rmi_device_read (self, addr++, 1, error);
+		if (f01_tmp == NULL) {
+			g_prefix_error (error, "failed to read query 42: ");
+			return FALSE;
+		}
+		priv->has_dds4_queries = (f01_tmp->data[0] & RMI_DEVICE_F01_QRY42_DS4_QUERIES) > 0;
+	}
+	if (priv->has_dds4_queries) {
+		g_autoptr(GByteArray) f01_tmp = NULL;
+		f01_tmp = fu_synaptics_rmi_device_read (self, addr++, 1, error);
+		if (f01_tmp == NULL) {
+			g_prefix_error (error, "failed to read DS4 query length: ");
+			return FALSE;
+		}
+		ds4_query_length = f01_tmp->data[0];
+	}
+	for (guint i = 1; i <= ds4_query_length; ++i) {
+		g_autoptr(GByteArray) f01_tmp = NULL;
+		f01_tmp = fu_synaptics_rmi_device_read (self, addr++, 1, error);
+		if (f01_tmp == NULL) {
+			g_prefix_error (error, "failed to read F01 Query43.%02x: ", i);
+			return FALSE;
+		}
+		switch (i) {
+		case 1:
```

Comment:
I think this is too heavy for a switch statement when there is just one case that matters.